File: 2_queue/RemovableQueue.ts
/* eslint-disable max-len */

interface IQueue<V> {
  readonly length: number
  push(value: V): void
  pop(): V | undefined
  top(): V | undefined
}

class RemovableQueue<V> {
  private readonly _queue: IQueue<V>
  private readonly _removedQueue: IQueue<V>
  private readonly _equals?: (v1: V, v2: V) => boolean

  constructor(createQueue: () => IQueue<V>, equals?: (v1: V, v2: V) => boolean) {
    this._queue = createQueue()
    this._removedQueue = createQueue()
    this._equals = equals
  }

  push(value: V): void {
    this._queue.push(value)
  }

  pop(): V | undefined {
    this._refresh()
    return this._queue.pop()
  }

  top(): V | undefined {
    this._refresh()
    return this._queue.top()
  }

  remove(value: V): void {
    this._removedQueue.push(value)
  }

  empty(): boolean {
    return this._removedQueue.length === 0
  }

  get length(): number {
    return this._queue.length - this._removedQueue.length
  }

  private _refresh(): void {
    if (!this._equals) {
      while (this._removedQueue.length && this._removedQueue.top() === this._queue.top()) {
        this._removedQueue.pop()
        this._queue.pop()
      }
    } else {
      while (this._removedQueue.length && this._equals(this._removedQueue.top()!, this._queue.top()!)) {
        this._removedQueue.pop()
        this._queue.pop()
      }
    }
  }
}

export {}

File: 2_queue/RemovableQueue.py
# 可删除元素的队列.(需要保证队列中不能有重复元素)

from collections import deque
from typing import Generic, Optional, Sequence, TypeVar


V = TypeVar("V")


class RemovableQueue(Generic[V]):
    __slots__ = ("_queue", "_removedQueue")

    def __init__(self, values: Optional[Sequence[V]] = None) -> None:
        self._queue = deque(values) if values is not None else deque()
        self._removedQueue = deque()

    def append(self, value: V) -> None:
        self._queue.append(value)

    def popLeft(self) -> V:
        self._refresh()
        return self._queue.popleft()

    def top(self) -> V:
        self._refresh()
        return self._queue[0]

    def remove(self, value: V) -> None:
        """删除前必须保证value存在于队列."""
        self._removedQueue.append(value)

    def empty(self) -> bool:
        return len(self) == 0

    def __len__(self) -> int:
        return len(self._queue) - len(self._removedQueue)

    def _refresh(self) -> None:
        while self._removedQueue and self._removedQueue[0] == self._queue[0]:
            self._removedQueue.popleft()
            self._queue.popleft()


if __name__ == "__main__":
    q = RemovableQueue([1, 2, 1, 3, 4, 5])
    q.remove(1)
    q.remove(3)
    q.remove(5)
    print(q.popLeft())
    print(q.popLeft())
